core/src/test/java/org/apache/iceberg/TestCustomValidations.java

```diff
@@ -0,0 +1,643 @@
+          CommitFailedException.class,
+          "Cannot commit changes based on stale metadata",
+          uncommitted::commit);
+    } else {
+      AssertHelpers.assertThrows(
+          "Should throw a ValidationException on commit due to concurrent update causing the given predicate to return false",
+          ValidationException.class,
+          failMessage,
+          uncommitted::commit);
+    }
+
+    assertFailure.accept(env);
+  }
+
+  private <E, T> void testModifyingTableInsideValidationThrowsException(
+      Supplier<E> setupEnv,
+      Function<E, PendingUpdate<T>> pendingUpdateSupplier,
+      Consumer<E> assertFailure)
+      throws Exception {
+    E env = setupTableAndEnv(setupEnv);
+
+    AssertHelpers.assertThrows(
+        "Any attempts to modify a table inside a validation should throw an exception",
+        java.lang.UnsupportedOperationException.class,
+        "Cannot modify a static table",
+        () -> {
+          PendingUpdate<T> pendingUpdate = pendingUpdateSupplier.apply(env);
+          pendingUpdate.validateCurrentTable(
+              currentTable -> {
+                // illegal action
+                currentTable.updateProperties().set("custom_watermark", "2").commit();
+                return true;
+              },
+              "Test failed.");
+          pendingUpdate.commit();
+        });
+
+    assertFailure.accept(env);
+  }
+
+  private <E, T> void testValidationBehaviours(
+      Supplier<E> setup,
+      Function<E, PendingUpdate<T>> pendingUpdateSupplier,
+      Consumer<E> assertSuccess,
+      Consumer<E> assertFailure)
+      throws Exception {
+    testValidationPasses(setup, pendingUpdateSupplier, assertSuccess);
+    testValidationFails(setup, pendingUpdateSupplier, assertFailure);
+    testFirstValidationFails(setup, pendingUpdateSupplier, assertFailure);
+    testSecondValidationFails(setup, pendingUpdateSupplier, assertFailure);
+    testValidationFailsDueToConcurrentCommit(setup, pendingUpdateSupplier, assertFailure);
+    testModifyingTableInsideValidationThrowsException(setup, pendingUpdateSupplier, assertFailure);
+  }
+
+  private <T> void testValidationBehaviours(
+      Supplier<PendingUpdate<T>> pendingUpdateSupplier,
+      Runnable assertSuccess,
+      Runnable assertFailure)
+      throws Exception {
+    testValidationBehaviours(
+        () -> null,
+        (__) -> pendingUpdateSupplier.get(),
+        (__) -> assertSuccess.run(),
+        (__) -> assertFailure.run());
+  }
+
+  @Test
+  public void testCherryPickOperation() throws Exception {
+    class Setup {
+      final long firstSnapshotId;
+      final long overwriteSnapshotId;
+
+      Setup(long firstSnapshotId, long overwriteSnapshotId) {
+        this.firstSnapshotId = firstSnapshotId;
+        this.overwriteSnapshotId = overwriteSnapshotId;
+      }
+    }
+
+    testValidationBehaviours(
```

Comment:
I understand that we don't want to repeat code, but reading through this test makes it quite difficult to reason about what's exactly being tested and expected (because we're just calling a bunch of lambdas).

I'm curious what others think, but I would vote for having clear and readable than completely de-duplicated code.

Reply:
I appreciate the politely-worded comment but I am going to push back a little here. 

I did not write de-duplicated code for the sake of de-duplicated code here. In general, I agree with having clear and readable tests over deduplicated code and in fact, I originally wrote the tests without any "deduplication" but quickly found that maintaining them was a massive burden. I "deduplicated" the code to ease this and didn't personally feel the readability was compromised that badly. 

I'll leave this as unresolved for. now to see if anyone else finds the tests extremely unreadable, in which case I'll make the requested change.

Reply:
@nastra I've gotten rid of all the lambdas and written separate tests for each case now 😄